mobile/src/main/java/org/openhab/habdroid/ui/OpenHABWidgetAdapter.java

```diff
@@ -59,918 +55,1053 @@
-    private ArrayList<VideoView> videoWidgetList;
-    private ArrayList<MySmartImageView> refreshImageList;
-    private ArrayList<MjpegStreamer> mjpegWidgetList;
+
+    public interface ItemClickListener {
+        boolean onItemClicked(OpenHABWidget item); // true -> select position
+        void onItemLongClicked(OpenHABWidget item);
+    }
+
+    private static final int TYPE_GENERICITEM = 0;
+    private static final int TYPE_FRAME = 1;
+    private static final int TYPE_GROUP = 2;
+    private static final int TYPE_SWITCH = 3;
+    private static final int TYPE_TEXT = 4;
+    private static final int TYPE_SLIDER = 5;
+    private static final int TYPE_IMAGE = 6;
+    private static final int TYPE_SELECTION = 7;
+    private static final int TYPE_SECTIONSWITCH = 8;
+    private static final int TYPE_ROLLERSHUTTER = 9;
+    private static final int TYPE_SETPOINT = 10;
+    private static final int TYPE_CHART = 11;
+    private static final int TYPE_VIDEO = 12;
+    private static final int TYPE_WEB = 13;
+    private static final int TYPE_COLOR = 14;
+    private static final int TYPE_VIDEO_MJPEG = 15;
+
+    private final ArrayList<OpenHABWidget> mItems = new ArrayList<>();
+    private final ConnectionInfo mConnection;
+    private final LayoutInflater mInflater;
     private MyAsyncHttpClient mAsyncHttpClient;
+    private ItemClickListener mItemClickListener;
     private @ColorInt int mPrimaryForegroundColor;
+    private String mChartTheme;
+    private int mSelectedPosition = -1;
+    private final boolean mSelectionEnabled;
+
+    public OpenHABWidgetAdapter(Context context, MyAsyncHttpClient httpClient,
+            String openHABBaseUrl, String openHABUsername, String openHABPassword,
+            ItemClickListener itemClickListener, boolean selectionEnabled) {
+        super();
 
-    public OpenHABWidgetAdapter(Context context, int resource,
-                                List<OpenHABWidget> objects) {
-        super(context, resource, objects);
-        // Initialize video view array
-        videoWidgetList = new ArrayList<VideoView>();
-        refreshImageList = new ArrayList<MySmartImageView>();
-        mjpegWidgetList = new ArrayList<MjpegStreamer>();
+        mInflater = LayoutInflater.from(context);
+
+        mAsyncHttpClient = httpClient;
+        mConnection = new ConnectionInfo(openHABBaseUrl, openHABUsername, openHABPassword);
+        mItemClickListener = itemClickListener;
+        mSelectionEnabled = selectionEnabled;
 
         TypedArray a = context.obtainStyledAttributes(new int[] {
-            R.attr.colorControlNormal
+            R.attr.colorControlNormal,
+            R.attr.chartTheme
         });
         mPrimaryForegroundColor = a.getColor(0, 0);
+        mChartTheme = a.getString(1);
+
         a.recycle();
     }
 
-    @SuppressWarnings("deprecation")
+    public void update(List<OpenHABWidget> widgets) {
+        mItems.clear();
+        mItems.addAll(widgets);
+        notifyDataSetChanged();
```

Comment:
This call is problem. notifyDataSetChanged(); does refresh all. In my PR there is
void reload(ArrayList<OpenHABWidget> oldItems, ArrayList<OpenHABWidget> newItems)
I had to implement requireRefresh on OpenHABWidget to support that.
I tested your branch and it does refresh my WebViews on page if something else refreshes.

Reply:
From the initial description of this PR:

> This allows more fine-grained change handling **later**

(emphasis added by me)

Having both RecyclerView conversion and update logic change is what bothered me in your PR, as (IMHO) it makes following the update logic changes harder, thus I believe the right thing to do is splitting those two items into separate PRs.

Reply:
Ok, starting position for me was to fix unnecessary refreshes, not refactoring.. Refactoring was required to achieve it.. IMO, this is blocking (at least for me) to use application. So I suggest to proceed. Logic can go in next PR, that is fine. But is required.